aten/src/ATen/native/cuda/DistanceKernel.cu

```diff
@@ -106,53 +106,50 @@ __global__ static void pdist_kernel_cuda_impl(scalar_t * result, const scalar_t
   // This shared memory is significantly larger than necessary, but the
   // assumption is that it's not a bottleneck, and this is simple
   __shared__ scalar_t shared[forward_threads];
-  int lane = threadIdx.y % warpSize;
-  int warp_id = threadIdx.y / warpSize;
+  int lane = threadIdx.x % warpSize;
+  int warp_id = threadIdx.x / warpSize;
   if (lane == 0) {
     shared[warp_id] = agg;
   }
   __syncthreads();
-  agg = (threadIdx.y < blockDim.y / warpSize) ? shared[lane] : 0.0;
+  agg = (threadIdx.x < blockDim.x / warpSize) ? shared[lane] : 0.0;
   if (warp_id == 0) {
     // Only reduce theads with nonzero data
-    for (int offset = blockDim.y / warpSize / 2; offset > 0; offset /= 2) {
+    for (int offset = blockDim.x / warpSize / 2; offset > 0; offset /= 2) {
       F::agg(agg, WARP_SHFL_DOWN(agg, offset));
     }
   }
-  if (threadIdx.y == 0) {
-    const int64_t combs = n * (n - 1) / 2;
-    result[l * combs + k] = F::finish(agg, p);
+  if (threadIdx.x == 0) {
+    result[k] = F::finish(agg, p);
   }
 }
 
 template <typename scalar_t, typename F>
-__global__ static void pdist_backward_kernel_cuda_impl(scalar_t * buffer, const scalar_t * grad, const scalar_t * self, const scalar_t * dist, const int64_t n, const int64_t m, const scalar_t p) {
-  const int l = blockIdx.x;
-  const int k = blockIdx.z * blockDim.z + threadIdx.z;
-  const int init = blockIdx.y * blockDim.y + threadIdx.y;
-  const int stride = blockDim.y * gridDim.y;
-  const int combs = n * (n - 1) / 2;
+__global__ static void pdist_backward_kernel_cuda_impl(scalar_t * buffer, const scalar_t * grad, const scalar_t * self, const scalar_t * dist, int64_t gs, const int64_t n, const int64_t m, const int64_t combs, const scalar_t p) {
+  const int k = blockIdx.y * blockDim.y + threadIdx.y;
+  const int init = blockIdx.x * blockDim.x + threadIdx.x;
+  const int stride = blockDim.x * gridDim.x;
 
   if (k >= combs) {
     return;
   }
 
-  float n2 = n - .5;
+  double n2 = n - .5;
   // The -1 accounts for floating point truncation issues
   int64_t i = static_cast<int64_t>((n2 - device_sqrt<scalar_t>(n2 * n2 - 2 * k - 1)));
```

Comment:
See previous comment

```suggestion
  int64_t i = static_cast<int64_t>((n2 - device_sqrt<double>(n2 * n2 - 2 * k - 1)));
```

Reply:
ah that makes sense, and probably what @ngimel meant above.